File: src/patterns/behavioral/state/state.go
package main

import "fmt"

// Context
type VimContext struct {
	currState IVimState
	content string

	defaultState IVimState
	visualState IVimState
	commandState IVimState
	insertState IVimState
}

func newVimContext() *VimContext {
	vimContext := &VimContext{}

	baseState := BaseState{vimContext: vimContext}
	
	defaultState := &DefaultState{baseState}
	visualState := &VisualState{baseState}
	commandState := &CommandState{baseState}
	insertState := &InsertState{baseState}

	vimContext.defaultState = defaultState
	vimContext.visualState = visualState
	vimContext.commandState = commandState
	vimContext.insertState = insertState
	
	vimContext.currState = defaultState
	vimContext.content = ""

	return vimContext
}

func (c *VimContext) SetState(state IVimState) {
	c.currState = state
}

func (c *VimContext) CommandLine() {
	c.currState.CommandLine()
}

func (c *VimContext) Visual()  {
	c.currState.Visual()
}

func (c *VimContext) Default() {
	c.currState.Default()
}

func (c *VimContext) Insert() {
	c.currState.Insert()
}

func (c *VimContext) WriteLetter(letter string){
	c.currState.WriteLetter(letter)
}

// State
type IVimState interface {
	CommandLine()
	Visual()
	Default()
	Insert()

	WriteLetter(letter string)
}

type BaseState struct {
	vimContext *VimContext
}

func (s *BaseState) CommandLine() {
	fmt.Println("Unable to go into the command mode from this state")
}

func (s *BaseState) Visual(){
	fmt.Println("Unable to go into the visual mode from this state")
}

func (s *BaseState) Insert(){
	fmt.Println("Unable to go into the insert mode from this state")
}

func (s *BaseState) Default(){
	s.vimContext.SetState(s.vimContext.defaultState)
}

func (s *BaseState) WriteLetter(letter string) {
	fmt.Println("Unable to write in the current mode")
}

type DefaultState struct {
	BaseState
}

func (s *DefaultState) CommandLine() {
	s.vimContext.SetState(s.vimContext.commandState)
}

func (s *DefaultState) Visual() {
	s.vimContext.SetState(s.vimContext.visualState)
}

func (s *DefaultState) Insert(){
	s.vimContext.SetState(s.vimContext.insertState)
}

func (s *DefaultState) Default() {
	fmt.Println("Already in that state")
}

type VisualState struct {
	BaseState
}

func (s *VisualState) Visual(){
	fmt.Println("Already in that state")
}

type InsertState struct {
	BaseState
}

func (s *InsertState) Insert(){
	fmt.Println("Already in that state")
}

func (s *InsertState) WriteLetter(letter string) {
	s.vimContext.content += letter
}

type CommandState struct {
	BaseState
}

func (s *CommandState) CommandLine(){
	fmt.Println("Already in that state")
}

// func main() {
// 	vim := newVimContext()
// 	vim.Insert()
// 	vim.WriteLetter("h")
// 	vim.WriteLetter("e")
// 	vim.WriteLetter("l")
// 	vim.WriteLetter("l")
// 	vim.WriteLetter("o")
// 	fmt.Println(vim.content)
// 	vim.Default()
// 	vim.Visual()
// }

File: src/patterns/behavioral/state/state.java
package practice;

// Context
interface ChangeState {
    void changeState(State state);
}

class Vim implements ChangeState, State{
    public State visualState;
    public State insertState;
    public State defaultState;
    public State commandState;
    private State currentState;

    public Vim() {
        defaultState = new DefaultState(this);
        visualState = new VisualState(this);
        insertState = new InsertState(this);
        commandState = new CommandState(this);
        currentState = defaultState;
    }

    public void changeState(State state) {
        currentState = state;
    }

    public void visualState() {
        currentState.visualState();
    }

    public void defaultState() {
        currentState.defaultState();
    }

    public void commandState() {
        currentState.commandState();
    }

    public void insertState() {
        currentState.insertState();
    }

    public void type() {
        currentState.type();
    }
}

// States
interface State {
    void visualState();
    void defaultState();
    void commandState();
    void insertState();
    void type();
}

abstract class BaseState implements State {
    protected Vim ctx;
    
    public BaseState(Vim ctx) {
        this.ctx = ctx;
    }

    public void visualState() {
        System.out.println("Switching to the visual state");
        ctx.changeState(ctx.visualState);
    }

    public void defaultState() {
        System.out.println("Switching to the default state");
        ctx.changeState(ctx.defaultState);
    }

    public void commandState() {
        System.out.println("Switching to the command state");
        ctx.changeState(ctx.commandState);
    }

    public void insertState() {
        System.out.println("Switching to the insert state");
        ctx.changeState(ctx.insertState);
    }

    public void type() {
        System.out.println("Unable to type in the current state");
    }
}

class InsertState extends BaseState {
    public InsertState(Vim ctx) { super(ctx); }

    public void insertState() {
        System.out.println("Already in the insert state");
    }

    public void type() {
        System.out.println("Typing...");
    }
}

class DefaultState extends BaseState {
    public DefaultState(Vim ctx) { super(ctx); }

    public void defaultState() {
        System.out.println("Already in the default state");
    }
}

class VisualState extends BaseState {
    public VisualState(Vim ctx) { super(ctx); }

    public void visualState() {
        System.out.println("Already in the default state");
    }
}

class CommandState extends BaseState {
    public CommandState(Vim ctx) { super(ctx); }

    public void commandState() {
        System.out.println("Already in the command state");
    }
}

// class Main {
//     public static void main(String[] args) {
//         Vim vim = new Vim();
//         vim.insertState();
//         vim.type();
//         vim.defaultState();
//         vim.defaultState();
//     }
// }